### ISSUE TITLE
[ReadSupport] Add support for reading from Neo4j by a provided query

### ISSUE THREAD

**Issue by conker84** [neo4j-contrib/neo4j-spark-connector#140]
We want to read data from Neo4j by allowing the developer to provide a simple Cypher query like:
```scala
val simpleDf = sparkSession
   .read
   .format("neo4j")
   .option("url", "neo4j://localhost:7687")
   .option("query", "MATCH (p:Person) RETURN p")
   .load()
```

In this case we should “flatten” the result as we already did in the Neo4j JDBC driver:
https://neo4j-contrib.github.io/neo4j-jdbc/#_flattening
and
https://github.com/neo4j-contrib/neo4j-jdbc/blob/4.0/neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltNeo4jResultSet.java#L154
The flattening level should be provided via option param.

Schema inferred automagically:
* Using APOC  “apoc.meta.nodeTypeProperties” and “apoc.meta.relTypeProperties”, we’ll make sure to provide access to APOC arguments via Spark options.
* If APOC are not present, we’ll use a prepared query like “MATCH (n:Person) RETURN n ORDER BY rand() LIMIT <flattening_value_provided_by_option_param>”

*N.b.* this feature will **not** support push-down filtering

**Comment by moxious** [comment]
Extra things required here:

We need a "mode" option which can be read/write, which should always default to "read".  This is important because in the underlying driver, we need to always use explicit transactions and not autocommit transactions.  But to do this we have to know if it's a read or write query.   Unfortunately explicit transactions in the underlying driver are super important to use, because it has big consequences on the routing at the driver level.

Your first example is good.  Here are extra examples to clarify desired behavior:

Examples:

This should work:

```
val simpleDf = sparkSession
   .read
   .format("neo4j")
   .option("url", "neo4j://localhost:7687")
   .option("mode", "write")
   .option("query", "CREATE (f:Foo) RETURN true as result")
   .load()
```

Without the `.option("mode","write")` this same thing should *fail*

This would also fail:  (probably because the driver will send it to a follower and you'll get a driver error)

```
val simpleDf = sparkSession
   .read
   .format("neo4j")
   .option("url", "neo4j://localhost:7687")
   .option("mode", "read")
   .option("query", "CREATE (f:Foo) RETURN true as result")
   .load()
```

But if you do a read-only query with mode "write" it should succeed.   This would be an important work-around to also force query routing to the leader if the user wants to do that.

**Comment by conker84** [comment]
closed via 7f1a9eb
